### PR TITLE
Dynamic target packet definition creation support. 

### LINKED ISSUE #419: Dynamically declared telemetry packets
During a recent training we discussed supporting dynamically declared packets. Conops is that various commands are sent to the target embedded software and the contents of a packet changes. COSMOS only updates packet definitions based on plugins so this would require building and installing a new plugin. However you can break up plugins however you want so you can create two plugins: one for the statically defined packets and one that changes based on target SW updates.

### LINKED ISSUE #1026: Cleanup other plugin versions on plugin install

### LINKED ISSUE #1044: Localmode plugins contain multiple folders for the same plugin after upgrading
I believe related to changes in plugin.txt

### PR DESCRIPTION
**Pull request by ryanmelt**
Also 
Improve build time to only rebuild docs tools on base change
Update docusaurus

closes #419 
closes #1026 
closes #1044

### CODE REVIEW

File: openc3/lib/openc3/packets/packet.rb [hunk 1]
```diff
@@ -872,6 +876,33 @@ def define_reserved_items
...
+      end
+      @sorted_items.each do |item|
+        if item.data_type == :DERIVED
+          new_sorted_items << item
+        end
+      end
+      @items = new_items
+      @sorted_items = new_sorted_items
+      clear_config_name()
+    end
```
Comment (jmthomas) [inline openc3/lib/openc3/packets/packet.rb:904]: I don't see this being called

File: openc3/lib/openc3/packets/packet.rb [hunk 2]
```diff
@@ -1019,6 +1050,13 @@ def to_config(cmd_or_tlm)
       else
         config << "COMMAND #{@target_name.to_s.quote_if_necessary} #{@packet_name.to_s.quote_if_necessary} #{@default_endianness} \"#{@description}\"\n"
       end
+      if @accessor.class.to_s != 'OpenC3::BinaryAccessor'
+        config << "  ACCESSOR #{@accessor.class.to_s} #{@accessor.args.map { |a| a.to_s.quote_if_necessary }.join(" ")}\n"
+      end
+      # TODO: Add TEMPLATE_ENCODED so this can always be done inline regardless of content
+      if @template
+        config << "  TEMPLATE '#{@template}'"
+      end
```
Comment (jmthomas) [inline openc3/lib/openc3/packets/packet.rb:1059]: I don't understand this TODO comment

File: openc3/lib/openc3/system/system.rb [hunk 3]
```diff
@@ -92,6 +92,18 @@ def self.setup_targets(target_names, base_dir, scope:)
...
+          # This adds support for remembering dynamically created packets
+          # target.txt must be configured to either use all files in cmd_tlm folder (default)
+          # or have a predetermined empty file like dynamic.txt
+          bucket_path = "#{scope}/targets_modified/#{target_name}/cmd_tlm"
+          dirs, files = bucket.list_files(bucket: ENV['OPENC3_CONFIG_BUCKET'], path: bucket_path)
+          files.each do |file|
+            bucket_key = File.join(bucket_path, file['name'])
+            local_path = "#{base_dir}/targets/#{target_name}/cmd_tlm/#{file['name']}"
+            bucket.get_object(bucket: ENV['OPENC3_CONFIG_BUCKET'], key: bucket_key, path: local_path)
+          end
```
Comment (jmthomas) [inline openc3/lib/openc3/system/system.rb:106]: This block is pulling down the dynamically created packets back into the local target_modified directory?

File: openc3/lib/openc3/models/target_model.rb [hunk 4]
```diff
@@ -794,23 +800,77 @@ def update_store(system)
...
+          client.put_object(
+            # Use targets_modified to save modifications
+            # This keeps the original target clean (read-only)
+            bucket: ENV['OPENC3_CONFIG_BUCKET'],
+            key: bucket_key,
+            body: config
+          )
+        end
+      end
+    end
```
Comment (jmthomas) [inline openc3/lib/openc3/models/target_model.rb:872]: How does this get called?

**Review by jmthomas — COMMENTED** [review]
Intriguing but I'm not sure how to use it

File: openc3/lib/openc3/models/plugin_model.rb [hunk 5]
```diff
@@ -158,6 +158,11 @@ def self.install_phase2(plugin_hash, scope:, gem_file_path: nil, validate_only:
           gem_file_path = OpenC3::GemModel.get(gem_name)
         end
 
+        # Attempt to remove all older versions of this same plugin before install to prevent version conflicts
+        # Especially on downgrades
+        # Leave the same version if it already exists
+        OpenC3::GemModel.destroy_all_other_versions(File.basename(gem_file_path))
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3/lib/openc3/models/plugin_model.rb:164]: The same version just gets left for the install.

File: openc3/lib/openc3/packets/packet.rb (hunk 1 above)
Comment (ryanmelt) [reply openc3/lib/openc3/packets/packet.rb:904]: Its called by the user.  See my prometheus plugin soon.

File: openc3/lib/openc3/packets/packet.rb (hunk 2 above)
Comment (ryanmelt) [reply openc3/lib/openc3/packets/packet.rb:1059]: I'll write the corresponding ticket.  In short we need a TEMPLATE keyword that takes base64 encoded data so a multiline string can be handled in one line of config file.

Comment (ryanmelt) [reply openc3/lib/openc3/packets/packet.rb:1059]: https://github.com/OpenC3/cosmos/issues/1062

File: openc3/lib/openc3/models/target_model.rb (hunk 4 above)
Comment (ryanmelt) [reply openc3/lib/openc3/models/target_model.rb:872]: Called by user.  See prometheus metrics plugin.

File: openc3/lib/openc3/system/system.rb (hunk 3 above)
Comment (ryanmelt) [reply openc3/lib/openc3/system/system.rb:106]: It is pulling them down into the files used to create the System object yes.

**Pull request merged**